transactionoutbox-cdi/src/main/java/com/gruelbox/transactionoutbox/CdiTransactionManager.java

```diff
@@ -2,173 +2,150 @@
-   public void inTransaction(TransactionalWork work)
-   {
-      uncheck(() -> inTransactionReturnsThrows(ThrowingTransactionalSupplier.fromWork(work)));
-   }
-
-   @Override
-   @Transactional(value = TxType.REQUIRES_NEW)
-   public <T, E extends Exception> T inTransactionReturnsThrows(ThrowingTransactionalSupplier<T, E> work) throws E
-   {
-      return work.doWork(transactionInstance);
-   }
-
-   @Override
-   public <T, E extends Exception> T requireTransactionReturns(ThrowingTransactionalSupplier<T, E> work) throws E, NoTransactionActiveException
-   {
-      if (tsr.getTransactionStatus() != Status.STATUS_ACTIVE)
-      {
-         throw new NoTransactionActiveException();
-      }
-
-      return work.doWork(transactionInstance);
-   }
-
-   private final class CdiTransaction implements Transaction
-   {
-
-      private DataSource datasource;
-
-      public CdiTransaction(DataSource datasource) throws SQLException
-      {
-         this.datasource = datasource;
-      }
-
-      @Override
-      public final Connection connection()
-      {
-         try
-         {
-            return datasource.getConnection();
-         }
-         catch (SQLException e)
-         {
-            throw new RuntimeException(e);
-         }
+/** Transaction manager which uses cdi and jta. */
+@ApplicationScoped
+public class CdiTransactionManager implements ThreadLocalContextTransactionManager {
+
+  private final CdiTransaction transactionInstance = new CdiTransaction();
+
+  private final DataSource datasource;
+
+  private final TransactionSynchronizationRegistry tsr;
+
+  @Inject
+  public CdiTransactionManager(AgroalDataSource datasource, TransactionSynchronizationRegistry tsr)
+      throws SQLException {
+    this.datasource = datasource;
+    this.tsr = tsr;
+  }
+
+  @Override
+  @Transactional(value = TxType.REQUIRES_NEW)
+  public void inTransaction(Runnable runnable) {
+    uncheck(() -> inTransactionReturnsThrows(ThrowingTransactionalSupplier.fromRunnable(runnable)));
+  }
+
+  @Override
+  @Transactional(value = TxType.REQUIRES_NEW)
+  public void inTransaction(TransactionalWork work) {
+    uncheck(() -> inTransactionReturnsThrows(ThrowingTransactionalSupplier.fromWork(work)));
+  }
+
+  @Override
+  @Transactional(value = TxType.REQUIRES_NEW)
+  public <T, E extends Exception> T inTransactionReturnsThrows(
+      ThrowingTransactionalSupplier<T, E> work) throws E {
+    return work.doWork(transactionInstance);
+  }
+
+  @Override
+  public <T, E extends Exception> T requireTransactionReturns(
+      ThrowingTransactionalSupplier<T, E> work) throws E, NoTransactionActiveException {
+    if (tsr.getTransactionStatus() != Status.STATUS_ACTIVE) {
+      throw new NoTransactionActiveException();
+    }
+
+    return work.doWork(transactionInstance);
+  }
+
+  private final class CdiTransaction implements Transaction {
+
+    public final Connection connection() {
+      try {
+        return datasource.getConnection();
```

Comment:
This is still not the connection used by the currently active transaction. It's requesting a new connection from the pool.

And there is no guarantee that a Spring Boot application will be using Agroal as the connection pool.

I've had a bit of a read and here's the answer: https://stackoverflow.com/questions/9642643/datasourceutils-getconnection-vs-datasource-getconnection

Just inject the normal `DataSource` and use `DataSourceUtils.getConnection(dataSource)`. That will return the currently active transaction's connection if one is active, or a fresh connection if it isn't.

To be sure that this definitely never creates a new connection, call `TransactionSynchronizationManaher.isActualTransactionActive()` first, and throw `IllegalStateException` if it returns false (a transaction should definitely be active at this point!)

Reply:
I can guarantee you that with this implementation and Quarkus, this gets the current active connection.

DataSourceUtils.getConnection(dataSource) is Spring dependent so I cannot / don't want to use it

Reply:
OK.

By the sound of things from this: https://quarkus.io/guides/datasource - what you're describing is not just Quarkus, but Quarkus with Agroal as the chosen connection pool _and_ Naranya for transaction management. It sounds like this is the default though.

Sounds like this is really a Quarkus integration rather than a "bare" CDI implementation, so I'll be happy with it as-is if the module is renamed to `transactionoutbox-quarkus` and `CdiTransactionManager` is renamed to `QuarkusTransactionManager`.  `CdiInstantiator` is named sensibly.

Reply:
ok I renamed the module